gprofiler/php.py

```diff
@@ -0,0 +1,234 @@
+import fcntl
+import glob
+import logging
+import os
+import re
+import signal
+from collections import Counter, defaultdict
+from functools import lru_cache
+from pathlib import Path
+from subprocess import Popen
+from threading import Event
+from typing import List, Mapping, MutableMapping, Optional, Pattern
+
+from gprofiler.exceptions import StopEventSetException
+from gprofiler.profiler_base import ProfilerBase
+from gprofiler.utils import limit_frequency, resource_path, start_process, wait_event
+
+logger = logging.getLogger(__name__)
+
+
+class PHPSpyProfiler(ProfilerBase):
+    PHPSPY_RESOURCE = "php/phpspy"
+    # Currently tracing only php-fpm, TODO: support mod_php in apache.
+    DEFAULT_PROCESS_FILTER = "php-fpm"
+    dump_signal = signal.SIGUSR2
+    poll_timeout = 10  # seconds
+    MAX_FREQUENCY = 999
+    MIN_DURATION = 3  # seconds, phpspy is running commands at bootstrap and it takes some time.
+    BUFFER_SIZE = 16384
+    NUM_WORKERS = 16  # Num of workers following unique PHP processes (one worker per process)
+
+    # regex required for parsing phpspy output
+    SINGLE_FRAME_RE = re.compile(r'^(?P<f_index>\d+) (?P<line>.*)$')  # "1 main.php:24"
+    PID_METADATA_RE = re.compile(r'^# pid = (.*)$')  # "# pid = 455"
+    PHP_FRAME_ANNOTATION = "[php]"
+
+    def __init__(
+        self,
+        frequency: int,
+        duration: int,
+        stop_event: Optional[Event],
+        storage_dir: str,
+        php_process_filter: str = DEFAULT_PROCESS_FILTER,
+    ):
+        self._frequency = limit_frequency(self.MAX_FREQUENCY, frequency, "phpspy", logger)
+
+        if duration < self.MIN_DURATION:
+            logger.warning(
+                f"Minimum duration for phpspy is {self.MIN_DURATION} (given {duration}), "
+                "raise duration in order to profile php processes"
+            )
+            self._enabled = False
```

Comment:
You fallthrough to `self._enabled = True`

Reply:
Why did you take this approach (of `_enabled`) instead of raising? I think it complicates things (for example, you can drop a `return` here, but then the `stop()` method which tries to access `self.process` will fail